File: misc/findtests/findtests.py
import os
import re
import fnmatch

def find_files_with_regex(directory, pattern):
    """
    Recursively find files matching the given regex pattern in a directory.
    """
    files = []
    regex = re.compile(pattern)
    for root, _, filenames in os.walk(directory):
        for filename in filenames:
            if regex.match(filename):
                files.append(os.path.join(root, filename))
    return files

def filter_files(file_list, exclusion_pattern):
    filtered_files = []
    for file in file_list:
        if not any(fnmatch.fnmatch(file, pattern) for pattern in exclusion_pattern):
            filtered_files.append(file)

    return filtered_files


def check_for_test_files(source_files, test_directory):
    """
    Check if test files exist for the given source files in the test directory.
    """
    filtered_files = []
    for source_file in source_files:
        base_name = os.path.splitext(os.path.basename(source_file))[0]
        possible_test_names = [f"{base_name}.test.ts", f"{base_name}.test.tsx"]

        found_test = False
        for root, _, filenames in os.walk(test_directory):
            if any(test_name in filenames for test_name in possible_test_names):
                found_test = True
                filtered_files.append(os.path.join(root, source_file))
                break

def print_file_list(file_list, message):
    print(f"'{message}':")
    count = 0
    for file in sorted(file_list):
        print(file)
        count += 1
    print(f"Total files found: {count}")

fws_directory = "/Users/abb2175/repos/work/fws/src"
bws_directory = "/Users/abb2175/repos/work/bws/src"
typescript_files = r"^(?!.*\.d\.tsx?$).*\.(ts|tsx)$"  # Regex to exclude .d.ts and .d.tsx files
exclusions = ["**/__generated__/*", "src/**/*.story.tsx", "src/client/routes/generated-routes.tsx", "src/server/index.tsx", "src/storybook-docs/**/*", "src/third-party/**/*", "src/**/*.mock-data.ts", "src/**/*.mocks.ts"]
test_files = r".*\.test\.tsx?$"
# sonar.exclusions=/node_modules/**/*, **/__generated__/*, src/**/*.story.tsx, src/client/routes/generated-routes.tsx, src/server/index.tsx, src/storybook-docs/**/*, src/third-party/**/*, src/**/*.mock-data.ts, src/**/*.mocks.ts
# sonar.test.inclusions=src/**/*.test.ts, src/**/*.test.tsx
# sonar.coverage.exclusions=src/**/*.story.tsx, src/storybook-docs/**/*, src/third-party/**/*, src/client/index.tsx, src/client/pages/dev-resources/**, **/jest.config.js, **/__mocks__/*, **/jest-*
# sonar.test.exclusions=src/**/*.story.tsx, src/storybook-docs/**/*, src/third-party/**/*, src/client/pages/dev-resources/**


def main():
    # find all typescript files in FWS
    source_files = find_files_with_regex(fws_directory, typescript_files)
    print_file_list(source_files, "Files found in FWS")
    # exclude files that match the exclusion patterns
    filtered_files = filter_files(source_files, exclusions)
    print_file_list(filtered_files, "Filtered files")
    # exclude files that have a matching test file
    no_test_files = check_for_test_files(source_files, fws_directory)
    print_file_list(no_test_files, "Files without test files")
    # print the remaining files in FWS
    # find all test files in BWS
    # find test files that match the remaining files

main()

# Check if corresponding test files exist in the test directory
# check_for_test_files(source_files, test_directory)

File: misc/findtests/findtests.ts
import fs from 'fs';
import path from 'path';
import { glob } from 'glob';
import { minimatch } from 'minimatch';

const findFiles = (
	pattern: string | string[],
	directory: string,
	exclusionPatterns?: string[]
) =>
	glob.globSync(pattern, {
		cwd: directory,
		ignore: exclusionPatterns || [],
		dot: true,
	});

const getFileName = (filePath: string) => path.basename(filePath);

// For each file in the sourceFiles array, check if there is a corresponding test file
// If there is no test file, add the file to a list of files without test files
const findTestFilesInSameDirectory = (
	sourceFiles: string[],
	testFiles: string[]
): string[] => {
	const noTestFiles: string[] = [];
	sourceFiles.forEach((sourceFile) => {
		const testFile = sourceFile
			.replace(/\.ts$/, '.test.ts')
			.replace(/\.tsx$/, '.test.tsx')
			.replace(/\.component\./, '.');
		if (!testFiles.includes(testFile)) {
			noTestFiles.push(sourceFile);
		}
	});
	return noTestFiles;
};

function findPossibleTestMatches(
	sourceFilesWithoutTests: string[],
	testFilesInBWS: string[]
) {
	const possibleTests: string[] = [];
	const sourceFilesWithoutPossibleTests: string[] = [];
	sourceFilesWithoutTests.forEach((sourceFile) => {
		const testFile = sourceFile
			.replace(/\.ts$/, '.test.ts')
			.replace(/\.tsx$/, '.test.tsx');
		const testFileBaseName = getFileName(testFile);
		const testFileInBWS = testFilesInBWS.find(
			(file) =>
				getFileName(file).replace(/\.component\./, '.') ===
				getFileName(testFile)
		);
		if (
			testFileInBWS?.includes('client/utils/string/general.utils.test.ts')
		) {
			// console.log('sourceFile', sourceFile);
		}
		if (testFileInBWS) {
			possibleTests.push(testFileInBWS);
		} else {
			sourceFilesWithoutPossibleTests.push(sourceFile);
		}
	});
	return [possibleTests, sourceFilesWithoutPossibleTests];
}

type PrintFileListOptions = {
	details: boolean;
};
const printFileList = (
	fileList: string[],
	message: string,
	options?: PrintFileListOptions
) => {
	console.log(`${message}: ${fileList.length}`);
	if (options?.details) {
		fileList.sort().forEach((file) => console.log(file));
	}
};

function main() {
	const fwsDirectory = '/Users/abb2175/repos/work/fws/src';
	const bwsDirectory = '/Users/abb2175/repos/work/bws/src';
	const typescriptFilePattern = '**/*.{ts,tsx}';
	const testFilePattern = '**/*.test.{ts,tsx}';
	const exclusionPatterns = [
		'**/*.d.{ts,tsx}',
		'**/__generated__/*',
		'**/*.story.tsx',
		'**/client/routes/generated-routes.tsx',
		'**/server/index.tsx',
		'**/client/index.tsx',
		'**/storybook-docs/**/*',
		'**/third-party/**/*',
		'**/**/*.mock-data.ts',
		'**/**/*.mocks.ts',
		'**/__mocks__/**',
		'**/jest.config.js',
		'**/jest',
		'**/graphql/*.package.ts',
		'**/graphql/root-package.ts',
	];
	const sourceFiles = findFiles(typescriptFilePattern, fwsDirectory, [
		...exclusionPatterns,
		testFilePattern,
	]);
	printFileList(sourceFiles, 'Typescript files found in FWS');

	const testFilesInFWS = findFiles(
		testFilePattern,
		fwsDirectory,
		exclusionPatterns
	);
	printFileList(testFilesInFWS, 'Test files found in FWS');

	const sourceFilesWithoutTests = findTestFilesInSameDirectory(
		sourceFiles,
		testFilesInFWS
	);
	printFileList(sourceFilesWithoutTests, 'Files without test files', {
		details: false,
	});

	const testFilesInBWS = findFiles(
		testFilePattern,
		bwsDirectory,
		exclusionPatterns
	);
	printFileList(testFilesInBWS, 'Test files found in BWS', {
		details: false,
	});

	const [possibleTests, sourceFilesWithoutPossibleTests] =
		findPossibleTestMatches(sourceFilesWithoutTests, testFilesInBWS);
	printFileList(possibleTests, 'Possible matching test files in BWS', {
		details: false,
	});
	printFileList(
		sourceFilesWithoutPossibleTests,
		'Files without possible matching test files',
		{
			details: true,
		}
	);
}

main();
